### DISCUSSION TITLE
Request for Simplified Custom `trace_id` Implementation in `CallbackHandler` for LangServe

### DISCUSSION THREAD

**Discussion by wertycn** [langfuse/langfuse#4330] [Support] [answered]
Hello everyone,

I'm currently working with `langfuse.callback.CallbackHandler` in LangServe for observation purposes. Due to specific requirements in my project, I need to customize the `trace_id`. However, I'm encountering difficulties in finding a straightforward method to achieve this.

Here's what I've tried so far:

1. **Using `langfuse_context._set_root_trace_id(trace_id=trace_id)`**: Unfortunately, this approach doesn't seem to work and leads to errors, although the program continues to run. The error I encountered is as follows:
   ```
   Traceback (most recent call last):
     File "xxxxx/py_deps_3_12_langfuse/site-packages/langfuse/client.py", line 1335, in trace
       new_body = TraceBody(**new_dict)
                  ^^^^^^^^^^^^^^^^^^^^^
     File "xxxx/py_deps_3_12_pydantic/site-packages/pydantic/v1/main.py", line 341, in __init__
       raise validation_error
   pydantic.v1.error_wrappers.ValidationError: 1 validation error for TraceBody
   tags -> 0
     none is not an allowed value (type=type_error.none.not_allowed)
   ```

2. **Specifying a `stateful_client` for `CallbackHandler`**: I tried setting a `stateful_client`, but it only seems to support `StatefulTraceClient` and `StatefulSpanClient`. To make this work, I had to introduce additional checks to verify if the trace was already created. While this solution is functional, it's more complex than I would like.

I am reaching out to ask if there is a simpler and more effective way to implement a custom `trace_id` in `CallbackHandler`. Any guidance or suggestions would be greatly appreciated.

Thank you for your assistance and looking forward to your feedback

**Comment by jannikmaierhoefer** [comment, marked as answer]
Hi @wertycn, you could use the Langfuse `@observe()` decorator together with the callback handler to pass the trace_id name: https://langfuse.com/docs/sdk/python/decorators#langchain

Here on how to pass the trace_id using the decorator: https://langfuse.com/docs/sdk/python/decorators#set-parent-trace-id-or-parent-span-id

Let me know if this would work in your case!
